Support for prepared statements

justinkluever opened the pull request:
How to use:

```lua
PulsarLib.SQL:PreparedQuery("SELECT 1 = ? WHERE stuff = ?", { 1 },
    function()
        print("ok")
    end
)
```

every ? in the query will get replaced by the value in the second argument (in the order it was given)
if there are more ? in the query than there are values in the second argument it will set everything that is missing to "NULL" in the query, this basicly translates "nil" to "NULL" for the database

the given example would result in this query:
`SELECT 1 = 1 WHERE stuff = NULL`


for sqlite it will use the fallback "prepareStatement" function to make it also work with sqlite, just without the actual prepared part

this merge request also fixes a small error related to logging

justinkluever committed:
feat: support for prepared statements

justinkluever committed:
fix: use PulsarLib instead of PulsarStore

justinkluever committed:
fix: use correct escape method

Lythium4848 commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -106,6 +138,61 @@ function SQL:RawQuery(query, onSuccess, onError)
 end
 
+function SQL:PreparedQuery(query, values, onSuccess, onError)
+	onSuiccess = onSuccess or emptyFunction
```
`onSuiccess` should be `onSuccess`

justinkluever replied:
Didn't noticed it first, but it will get fixed ^^

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -56,6 +56,38 @@ function SQL:Connect()
+		elseif type(value) == "boolean" then
+			newQuery = newQuery .. (value and "1" or "0")
+		else
+			newQuery = newQuery .. value
```
If a player, panel or entity is passed, will it stringify when concated? If so, it'll be an unescaped, unquoted string.

justinkluever replied:
from what i can see it will throw errors on anything that doesnt support __concat 

i will  implement some checks tho

justinkluever replied:
i dont know why github doesnt detect this as changed, but it is changed already

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -56,6 +56,38 @@ function SQL:Connect()
+		if value == nil then
+			newQuery = newQuery .. "NULL"
+		elseif type(value) == "string" then
+			newQuery = newQuery .. self:Escape(value)
```
Not sure if self:Escape wraps in quotes? Could this be checked please.

justinkluever replied:
No it doesnt, i fixed SQL:Escape to do the same in either sqlite or mysql settings

JoshPiper replied:
If it's not wrapped, it'll just chuck a string without quotes into the query then? Probably needs a fix for that.

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -92,7 +124,7 @@ function SQL:RawQuery(query, onSuccess, onError)
 		local x = string.Split(query, "\n")
 		for _, line in ipairs(x) do
-			PulsarStore.Logging.Debug(line)
+			PulsarLib.Logging.Debug(line)
```
Good catch.

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -106,6 +138,61 @@ function SQL:RawQuery(query, onSuccess, onError)
+			onError(err)
+		end
+
+		for k, v in pairs(values or {}) do
```
ipairs?

justinkluever replied:
Yes

justinkluever committed:
fix: typo

justinkluever committed:
fix: use ipairs instead of pairs, should not be an issue

justinkluever committed:
fix: wrap Escape output in qoutes for mysql escape

Reason (with mysql enabled):
```
> print(PulsarLib.SQL:Escape("im a string '\""))...
im a string \'\"
```

this should not be an issue with mysql but just in case
(this would not get used unless for error printing in prepared statements unless sqlite is used)

justinkluever committed:
fix: assert number as last resort, bail if not

if it is neither nil, string, boolean or a number the developer put something
weird into it, lets prevent that from causing damage.

Lythium4848 commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -75,8 +75,11 @@ function SQL:prepareStatement(query, values)
+		elseif tonumber(value) then
 			newQuery = newQuery .. value
+		else
+			ErrorNoHaltWithStack("Invalid value type for prepared statement, expected nil, string, boolean or number, got " .. type(value))
```
You should use PulsarLib.Logging.Fatal here

JoshPiper replied:
.Fatal won't drop the stack, so I'd make two calls, one with the error, and one with debug.traceback() or whatever that function was.

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -92,8 +127,63 @@ function SQL:RawQuery(query, onSuccess, onError)
+		end
+
+		queryObj.onError = function(_, err)
+			PulsarLib.Logging.Fatal("Prepared MySQL query failed!")
+			PulsarLib.Logging.Fatal(err)
+			PulsarLib.Logging.Fatal(SQL:prepareStatement(query, values))
```
Can these calls be replaced with `local logger = PulsarLib.Logging:Get("Database")` -> `logger.Fatal(whatever)` please?

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -92,8 +127,63 @@ function SQL:RawQuery(query, onSuccess, onError)
+
+		local x = string.Split(query, "\n")
+		for _, line in ipairs(x) do
+			PulsarLib.Logging.Debug(line)
```
This should probably be at one of the trace levels.

justinkluever replied:
i don't understand what you mean with that, Debug is already very low and there are no Trace methods

JoshPiper replied:
Oh, is this the build with trace level, but no trace methods? I'll fix that later.

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -2,7 +2,7 @@ PulsarLib = PulsarLib or {}
 	stored = {}
 }, {__index = PulsarLib})
-
+ 
```
me when whitespace

JoshPiper commented on lua/pulsar_lib/core/sql/sv_sql.lua:
```diff
@@ -44,7 +44,7 @@ end
 function SQL:ConnectSQLite()
 	hook.Run("PulsarLib.SQL.Connected")
-end
+end 
```
me when whitespace 2

justinkluever committed:
fix: use PulsarLib.Logging.Fatal instead of gmod error stuff

justinkluever committed:
fix: pushed wrong copy oops

justinkluever committed:
fix: remove whitestpaces

justinkluever committed:
fix: use Database logger for database stuff

Lythium4848 reviewed (approved):
Looks good to me 🗿

JoshPiper reviewed (approved):
lgtm
